### PR TITLE
Split runtime components out of config in the orchestrator impl

### PR DESCRIPTION
**Pull request by jdisanti**
This PR moves all the "runtime components", pieces that are core to the operation of the orchestrator, into a separate `RuntimeComponents` type for the orchestrator to reference directly.

The reason for this is so that these core components cannot be changed by interceptors while the orchestrator is executing a request.

----

_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

### CODE REVIEW

File: codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/ClientCodegenVisitor.kt [hunk 1]
```diff
@@ -170,7 +170,8 @@ class ClientCodegenVisitor(
             ),
         )
         try {
-            "cargo fmt".runCommand(fileManifest.baseDir, timeout = settings.codegenConfig.formatTimeoutSeconds.toLong())
+            // use an increased max_width to make rustfmt fail less frequently
+            "cargo fmt -- --config max_width=150".runCommand(fileManifest.baseDir, timeout = settings.codegenConfig.formatTimeoutSeconds.toLong())
```
Comment (jdisanti) [inline codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/ClientCodegenVisitor.kt:174]: This should get split out into a separate PR to keep the codegen diff clean.

File: rust-runtime/aws-smithy-runtime-api/src/client/runtime_components.rs [hunk 2]
```diff
@@ -0,0 +1,732 @@
...
+            return Err(BuildError(concat!(
+                "at least one `",
+                stringify!($name),
+                "` runtime component is required"
+            )));
+        }
+        $self.$name
+    }};
+}
+macro_rules! rc_type {
```
Comment (rcoh) [inline rust-runtime/aws-smithy-runtime-api/src/client/runtime_components.rs:56]: maybe write this out? I thought this had something to do with rc

File: codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/ClientCodegenVisitor.kt (hunk 1 above)
Comment (jdisanti) [reply codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/ClientCodegenVisitor.kt:174]: Split out in https://github.com/awslabs/smithy-rs/pull/2835

**Review by ysaito1001 — COMMENTED** [review]
Fantastic work and thanks for the effort. Left minor comments just for clarification.

One potential concern is the amount of macros we now have to parse mentally when reading code, but as far as I can tell, their use is restricted to places that repeat some code patterns mechanically.

File: rust-runtime/aws-smithy-runtime-api/src/client/runtime_components.rs [hunk 3]
```diff
@@ -0,0 +1,732 @@
...
+}
+
+impl GetIdentityResolver for RuntimeComponents {
+    fn identity_resolver(&self, scheme_id: AuthSchemeId) -> Option<SharedIdentityResolver> {
+        self.identity_resolvers
+            .iter()
+            .find(|s| s.value.scheme_id() == scheme_id)
+            .map(|s| s.value.identity_resolver())
+    }
+}
```
[the remark this answers is not in the document]
Comment (jdisanti) [reply rust-runtime/aws-smithy-runtime-api/src/client/runtime_components.rs:597]: This is to ensure the trait for `HttpAuthScheme`'s `identity_resolver` method has the correct data available in the function signature. The internal spec states that it should only have access to identity resolvers, and not to everything that's in `RuntimeComponents`.

File: rust-runtime/aws-smithy-runtime/src/client/config_override.rs
[the remark this answers is not in the document]
Comment (jdisanti) [reply rust-runtime/aws-smithy-runtime/src/client/config_override.rs:1]: `override` is a reserved word in Rust, so I can't use that unfortunately.

[the remark this answers is not in the document]
Comment (ysaito1001) [reply rust-runtime/aws-smithy-runtime/src/client/config_override.rs:1]: Oh I didn't know! It makes sense, then.

**Comment by jdisanti** [comment]
> One potential concern is the amount of macros we now have to parse mentally when reading code, but as far as I can tell, their use is restricted to places that repeat some code patterns mechanically.

Yeah. Maybe I should add some comments to help with that.

File: rust-runtime/aws-smithy-runtime-api/src/client/runtime_components.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ysaito1001) [reply rust-runtime/aws-smithy-runtime-api/src/client/runtime_components.rs:597]: Thanks for clarifying! Can we add a doc based on what you explained to me (it's a `pub` item that deserves a doc anyway, and it helps others who are not familiar with our internal spec when they read it)?

File: rust-runtime/aws-smithy-runtime-api/src/client/runtime_plugin.rs [hunk 4]
```diff
@@ -19,8 +22,8 @@ pub trait RuntimePlugin: Debug + Send + Sync {
         None
     }
 
-    fn interceptors(&self, interceptors: &mut InterceptorRegistrar) {
-        let _ = interceptors;
+    fn runtime_components(&self) -> Cow<'_, RuntimeComponentsBuilder> {
```
[the remark this answers is not in the document]
Comment (jdisanti) [reply rust-runtime/aws-smithy-runtime-api/src/client/runtime_plugin.rs:25]: Updated in 435ab59252e88e73358933c3dc5004d3f15beb78

File: rust-runtime/aws-smithy-runtime-api/src/client/runtime_components.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jdisanti) [reply rust-runtime/aws-smithy-runtime-api/src/client/runtime_components.rs:650]: Addressed in e42c7aaa2da7519efc9836e1bec53618327533bc

File: rust-runtime/aws-smithy-runtime-api/src/client/runtime_components.rs (hunk 2 above)
Comment (jdisanti) [reply rust-runtime/aws-smithy-runtime-api/src/client/runtime_components.rs]: Addressed in 68e834d622a4a77afb388f8cc6cef81cc299cc4a

**Review by rcoh — APPROVED** [review]

File: codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/generators/client/CustomizableOperationGenerator.kt [hunk 5]
```diff
@@ -202,6 +204,13 @@ class CustomizableOperationGenerator(
                             self
                         }
 
+                        /// Adds a runtime plugin.
+                        ##[doc(hidden)]
+                        pub fn runtime_plugin(mut self, runtime_plugin: impl #{RuntimePlugin} + 'static) -> Self {
```
Comment (rcoh) [inline codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/generators/client/CustomizableOperationGenerator.kt:209]: does this really need to be pub? I kinda imagine pub(crate) would suffice. I also might name it `with_plugin(...)` to clarify

File: rust-runtime/aws-smithy-runtime-api/src/client/runtime_plugin.rs [hunk 6]
```diff
@@ -99,6 +99,41 @@ impl RuntimePlugins {
     }
 }
 
+#[derive(Default, Debug)]
+pub struct PassthroughRuntimePlugin {
+    config: Option<FrozenLayer>,
+    runtime_components: Option<RuntimeComponentsBuilder>,
+}
```
Comment (rcoh) [inline rust-runtime/aws-smithy-runtime-api/src/client/runtime_plugin.rs:106]: `RuntimePluginBuilder` maybe? `StaticRuntimePlugin`

File: codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/generators/config/ServiceConfigGenerator.kt [hunk 7]
```diff
@@ -380,8 +382,9 @@ class ServiceConfigGenerator(
             if (runtimeMode.defaultToOrchestrator) {
                 rustTemplate(
                     """
-                    config: #{CloneableLayer},
-                    runtime_components: #{RuntimeComponentsBuilder},
+                    pub(crate) config: #{CloneableLayer},
+                    pub(crate) runtime_components: #{RuntimeComponentsBuilder},
+                    pub(crate) runtime_plugins: #{Vec}<#{SharedRuntimePlugin}>,
```
Comment (rcoh) [inline codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/generators/config/ServiceConfigGenerator.kt:387]: are these custom/extra runtime plugins or the base ones?

File: codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/generators/config/ServiceConfigGenerator.kt [hunk 8]
```diff
@@ -412,6 +415,27 @@ class ServiceConfigGenerator(
...
+                    /// Adds a runtime plugin to the config.
+                    ##[doc(hidden)]
+                    pub fn runtime_plugin(mut self, plugin: impl #{RuntimePlugin} + 'static) -> Self {
+                        self.runtime_plugins.push(#{SharedRuntimePlugin}::new(plugin));
+                        self
+                    }
+
+                    /// Adds a runtime plugin to the config.
+                    ##[doc(hidden)]
+                    pub fn push_runtime_plugin(&mut self, plugin: #{SharedRuntimePlugin}) -> &mut Self {
```
Comment (rcoh) [inline codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/generators/config/ServiceConfigGenerator.kt:430]: I think this can be pub(crate)

File: codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/generators/config/ServiceConfigGenerator.kt [hunk 9]
```diff
@@ -412,6 +415,27 @@ class ServiceConfigGenerator(
                 it.section(ServiceConfig.BuilderImpl)(this)
             }
 
+            if (runtimeMode.defaultToOrchestrator) {
+                rustTemplate(
+                    """
+                    /// Adds a runtime plugin to the config.
+                    ##[doc(hidden)]
+                    pub fn runtime_plugin(mut self, plugin: impl #{RuntimePlugin} + 'static) -> Self {
+                        self.runtime_plugins.push(#{SharedRuntimePlugin}::new(plugin));
```
Comment (rcoh) [inline codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/generators/config/ServiceConfigGenerator.kt:424]: this can use `push_runtime_plugin`

**Review by rcoh — APPROVED** [review]
LGTM—but I think a lot of the doc hidden items can actually be made pub(crate)

File: codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/generators/config/ServiceConfigGenerator.kt (hunk 7 above)
Comment (jdisanti) [reply codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/generators/config/ServiceConfigGenerator.kt:387]: These are custom/extra.

File: codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/generators/client/CustomizableOperationGenerator.kt (hunk 5 above)
Comment (jdisanti) [reply codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/generators/client/CustomizableOperationGenerator.kt:209]: I kind of broke our builder convention with the `with_` methods, but I figured that's OK since it's for something more internal rather than the frequently used public facing API. Now that a lot of our public builders have getters though, we may consider bringing in the `with_` convention rather than having `thing`, `set_thing`, and `get_thing`.

File: codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/generators/config/ServiceConfigGenerator.kt (hunk 8 above)
Comment (jdisanti) [reply codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/generators/config/ServiceConfigGenerator.kt:430]: I seem to be running into a compiler bug where rustc erroneously gives a dead code warning even with `#[allow(unused)]` or `#[allow(dead_code)]` on top of the function. Going to stick with the pub/doc-hidden combo for now.

File: rust-runtime/aws-smithy-runtime-api/src/client/runtime_plugin.rs (hunk 6 above)
Comment (jdisanti) [reply rust-runtime/aws-smithy-runtime-api/src/client/runtime_plugin.rs]: Renamed in cacf7bac9993f8785ba087aa185e4fb69d5c1419

File: codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/generators/config/ServiceConfigGenerator.kt (hunk 8 above)
Comment (jdisanti) [reply codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/generators/config/ServiceConfigGenerator.kt]: Changed to `pub(crate)` in cacf7bac9993f8785ba087aa185e4fb69d5c1419

File: codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/generators/config/ServiceConfigGenerator.kt (hunk 9 above)
Comment (jdisanti) [reply codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/generators/config/ServiceConfigGenerator.kt]: Fixed in cacf7bac9993f8785ba087aa185e4fb69d5c1419

File: codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/generators/client/CustomizableOperationGenerator.kt (hunk 5 above)
Comment (jdisanti) [reply codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/generators/client/CustomizableOperationGenerator.kt]: Made `pub(crate)` in cacf7bac9993f8785ba087aa185e4fb69d5c1419

**Review by ysaito1001 — APPROVED** [review]
